filament/src/ResourceAllocator.cpp

```diff
@@ -254,4 +265,45 @@ void ResourceAllocator::purge(
     mTextureCache.erase(pos);
 }
 
+// ------------------------------------------------------------------------------------------------
+
+ResourceAllocatorDisposer::ResourceAllocatorDisposer(DriverApi& driverApi) noexcept
+        : mBackend(driverApi) {
+}
+
+ResourceAllocatorDisposer::~ResourceAllocatorDisposer() noexcept {
+     assert_invariant(mInUseTextures.empty());
+}
+
+void ResourceAllocatorDisposer::terminate() noexcept {
+    assert_invariant(mInUseTextures.empty());
+}
+
+void ResourceAllocatorDisposer::destroy(backend::TextureHandle handle) noexcept {
+    auto r = checkin(handle);
+    if (r.has_value()) {
+        mBackend.destroyTexture(handle);
+    }
+}
+
+void ResourceAllocatorDisposer::checkout(backend::TextureHandle handle,
+        ResourceAllocator::TextureKey key) {
+    mInUseTextures.emplace(handle, key);
+}
+
+std::optional<ResourceAllocator::TextureKey> ResourceAllocatorDisposer::checkin(
+        backend::TextureHandle handle) {
+    std::optional<ResourceAllocator::TextureKey> result;
+    // find the texture in the in-use list (it must be there!)
+    auto it = mInUseTextures.find(handle);
+    assert_invariant(it != mInUseTextures.end());
+    if (it == mInUseTextures.end()) {
+        return result;
+    }
```

Comment:
I'm a bit confused by this. It reads like in debug, we expect the handle to be in the in-use list, but in release, we don't?

Reply:
In debug we assert if we don't find the item in the list, and in release we return `result` which is a `std::optional<>` with no value set. But I will change it to `std::nullopt` to be clearer.

Reply:
sorry, to be clearer, why do we expect different behavior for debug and release?

Reply:
Well in both case we expect it to be in the list. But in DEBUG I assert (so we catch it) and in RELEASE I return an error so that we don't corrupt the state. I could remove the 
```
    if (it == mInUseTextures.end()) {
        return std::nullopt;
    }
```
because it should never happen.